### PR TITLE
Add msisensor (#95)

### PR DESCRIPTION
**Pull request by davidmasp**
# nf-core/sarek pull request

## Description
Adding MSIsensor as a tool to Sarek. (from here #95 )

First nf-core commit so I am a bit clueless of what I am doing.  Basically I added code to the pipeline to run [msisensor](https://github.com/ding-lab/msisensor) which detects MSI status from NGS tumor/normal pairs. It's divided in 2 steps which I would say fit very nicely in Sarek. I guess this is a WIP PR. Will update further but thought it would be better to submit already.

## PR checklist

- [x] This comment contains a description of changes (with reason)
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Make sure your code lints (`nf-core lint .`).
- [x] Documentation in `docs` is updated
- [x] `CHANGELOG.md` is updated
- [x] `README.md` is updated

**Learn more about contributing:** [CONTRIBUTING.md](https://github.com/nf-core/sarek/tree/master/.github/CONTRIBUTING.md)

### CODE REVIEW

**Comment by davidmasp** [comment]
I did build the environment in conda and I managed to run the `nextflow run . -profile test --tools msisensor -resume` and it does run but I don't have docker installed rn. I could try to configure it when I have more time though.

**Comment by davidmasp** [comment]
For the test I am not sure what to do, should I just include a test profile like `conf/test_msi.config` and there add tools msi? I mean, does it also check the output or just if the pipeline runs. 

Regarding the output, the tool runs but it detects 0% which makes sense because the test dataset is small so I was also wondering if I need to build a test dataset to get a "supervised" output or just the fact that it runs is enough? I am not sure how I would add microsattellite inestability in a bam file, I guess it's possible though.

File: main.nf [hunk 1]
```diff
@@ -2070,7 +2071,7 @@ pairBam = bamNormal.cross(bamTumor).map {
 pairBam = pairBam.dump(tag:'BAM Somatic Pair')
 
 // Manta, Strelka, Mutect2
```
Comment (MaxUlysse) [inline main.nf]:
```suggestion
// Manta, Strelka, Mutect2, MSIsensor
```

File: main.nf [hunk 2]
```diff
@@ -618,6 +618,7 @@ process Get_software_versions {
     trim_galore -v &> v_trim_galore.txt 2>&1 || true
     vcftools --version &> v_vcftools.txt 2>&1 || true
     vep --help &> v_vep.txt 2>&1 || true
+    msisensor &> v_msisensor.txt 2>&1 || true
```
Comment (MaxUlysse) [inline main.nf]: Can you also update the regexes in `bin/scrape_software_versions.py` to match the msisensor output?
https://github.com/nf-core/sarek/blob/52bee464a89b1a891d608637b67f5c9a1da8aead/bin/scrape_software_versions.py#L6

**Comment by MaxUlysse** [comment]
> For the test I am not sure what to do, should I just include a test profile like `conf/test_msi.config` and there add tools msi? I mean, does it also check the output or just if the pipeline runs.
> 
> Regarding the output, the tool runs but it detects 0% which makes sense because the test dataset is small so I was also wondering if I need to build a test dataset to get a "supervised" output or just the fact that it runs is enough? I am not sure how I would add microsattellite inestability in a bam file, I guess it's possible though.

For now, let's just try adding MSIsensor to the list here:
https://github.com/nf-core/sarek/blob/52bee464a89b1a891d608637b67f5c9a1da8aead/.github/workflows/ci.yml#L134
We'll figure something out later.
We have an issue here to make some test data for Mutect2:
https://github.com/nf-core/sarek/issues/148
I'm hoping that would be a solution for MSIsensor testing as well.

File: environment.yml [hunk 3]
```diff
@@ -21,6 +21,7 @@ dependencies:
   - bioconda::genesplicer=1.0
   - bioconda::htslib=1.9
   - bioconda::manta=1.6.0
+  - bioconda::msisensor=0.5
```
Comment (MaxUlysse) [inline environment.yml]: Could you update the `docs/containers.md` as well please?

File: main.nf [hunk 4]
```diff
@@ -2605,6 +2606,64 @@ process StrelkaBP {
...
+    tag {idSampleTumor + "_vs_" + idSampleNormal}
+
+    publishDir "${params.outdir}/MSI/${idSampleTumor}_vs_${idSampleNormal}/msisensor", mode: params.publishDirMode
+
+    input:
+    set idPatient, idSampleNormal, file(bamNormal), file(baiNormal), idSampleTumor, file(bamTumor), file(baiTumor) from pairBamMsisensor
+    file msiSites from msi_scan_ch
+
+    output:
+    set val("Msisensor"), idPatient, file("${idSampleTumor}_vs_${idSampleNormal}_msisensor"), file("${idSampleTumor}_vs_${idSampleNormal}_msisensor_dis"), file("${idSampleTumor}_vs_${idSampleNormal}_msisensor_germline"), file("${idSampleTumor}_vs_${idSampleNormal}_msisensor_somatic") into msisensor_out_ch
```
Comment (MaxUlysse) [inline main.nf]: Can you please add some information about the outputs in the docs?
https://github.com/nf-core/sarek/blob/dev/docs/output.md

File: environment.yml (hunk 3 above)
Comment (MaxUlysse) [reply environment.yml]: And while your at it, could you please also add a line in the CHANGELOG about adding msisensor:
https://github.com/nf-core/sarek/blob/dev/CHANGELOG.md

**Comment by davidmasp** [comment]
Okay, I think mostly I understand what i need to fill up

**Comment by davidmasp** [comment]
Will update when I have time, thanks!

File: environment.yml (hunk 3 above)
Comment (davidmasp) [reply environment.yml]: done in 41154f551aacaf2ddec2797b70839b0aad087dd4

**Comment by davidmasp** [comment]
> For now, let's just try adding MSIsensor to the list here:
> https://github.com/nf-core/sarek/blob/52bee464a89b1a891d608637b67f5c9a1da8aead/.github/workflows/ci.yml#L134
> 

okay, done in 9afaa94e345c6ed0c824919b9ebb7e61f050fe09

File: main.nf (hunk 2 above)
Comment (davidmasp) [reply main.nf]: done in e0af15e

**Comment by davidmasp** [comment]
How should I update the readme?

**Comment by MaxUlysse** [comment]
> How should I update the readme?

Add your name to the list of helpful contributors ;-)

**Comment by MaxUlysse** [comment]
But I guess you were speaking about the docs, and not the actual README file.

Basically, I would add a section in Variant Calling for microsatellites.
And then as we do for any other tool, just describe the actual tool (usually, I copy from their readme), and the output files.

**Comment by davidmasp** [comment]
Ah okok!

**Comment by davidmasp** [comment]
so only docs missing, I wanna check what I write is helpful and true so I'll take some time, not much I guess though.

File: main.nf (hunk 4 above)
Comment (davidmasp) [reply main.nf]: done in c1af38b97335644aaec2f2184b6463ac2d87f149

**Comment by davidmasp** [comment]
everything should be more or less done I think just lmk otherwise. The test isn´t passing because it´s not using the new docker built right? I am still not completely sure how this works.

Cheers

File: docs/output.md [hunk 5]
```diff
@@ -34,6 +34,8 @@ The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes d
     - [ConvertAlleleCounts](#convertallelecounts)
     - [ASCAT](#ascat)
     - [Control-FREEC](#control-freec)
+  - [MSI status](#msi-status)
+    - [msisensor](#msisensor)
```
Comment (MaxUlysse) [inline docs/output.md]:
```suggestion
    - [MSIsensor](#msisensor)
```

File: docs/output.md [hunk 6]
```diff
@@ -424,6 +426,36 @@ For a Tumor/Normal pair only:
...
   - file with beta allele frequencies for each possibly heterozygous SNP position
 
+### MSI status
+
+[Microsatellite instability](https://en.wikipedia.org/wiki/Microsatellite_instability)
+is a genetic condition associated to deficienceies in the
+mismatch repair (MMR) system which causes a tendency to accumulate a high
+number of mutations (SNVs and indels).
+
+#### msisensor
```
Comment (MaxUlysse) [inline docs/output.md]:
```suggestion
#### MSIsensor
```

File: docs/output.md [hunk 7]
```diff
@@ -424,6 +426,36 @@ For a Tumor/Normal pair only:
...
+
+[msisensor](https://github.com/ding-lab/msisensor) is a tool to detect the MSI
```
Comment (MaxUlysse) [inline docs/output.md]:
```suggestion
[MSIsensor](https://github.com/ding-lab/msisensor) is a tool to detect the MSI
```

File: docs/output.md [hunk 8]
```diff
@@ -424,6 +426,36 @@ For a Tumor/Normal pair only:
...
+status of a tumor scaning the length of the microsatellite regions. An altered
+distribution of  microsatellite length is associated to a missed replication
+slippage which would be corrected under normal MMR conditions. It requires
```
Comment (MaxUlysse) [inline docs/output.md]: What is MMR standing for?

File: docs/output.md [hunk 9]
```diff
@@ -424,6 +426,36 @@ For a Tumor/Normal pair only:
...
+a normal sample for each tumour to differentiate the somatic and germline
+cases.
+
+For further reading see the [msisensor paper](https://www.ncbi.nlm.nih.gov/pubmed/24371154).
```
Comment (MaxUlysse) [inline docs/output.md]:
```suggestion
For further reading see the [MSIsensor paper](https://www.ncbi.nlm.nih.gov/pubmed/24371154).
```

File: main.nf [hunk 10]
```diff
@@ -2605,6 +2606,64 @@ process StrelkaBP {
...
+// Score the normal vs somatic pair of bams
+
+process msisensor {
+    label 'cpus_4'
+    label 'memory_max'
+    // memory '10 GB'
+
+    tag {idSampleTumor + "_vs_" + idSampleNormal}
+
+    publishDir "${params.outdir}/MSI/${idSampleTumor}_vs_${idSampleNormal}/msisensor", mode: params.publishDirMode
```
Comment (MaxUlysse) [inline main.nf]:
```suggestion
    publishDir "${params.outdir}/MSIsensor/${idSampleTumor}_vs_${idSampleNormal}/msisensor", mode: params.publishDirMode
```

File: docs/output.md [hunk 11]
```diff
@@ -424,6 +426,36 @@ For a Tumor/Normal pair only:
...
+
+For a Tumor/Normal pair only:
+**Output directory: `results/MSI/[TUMORSAMPLE]_vs_[NORMALSAMPLE]/msisensor`**
```
Comment (MaxUlysse) [inline docs/output.md]:
```suggestion
**Output directory: `results/MSIsensor/[TUMORSAMPLE]_vs_[NORMALSAMPLE]/msisensor`**
```

**Comment by MaxUlysse** [comment]
That looks perfect, I just made a couple of more suggestions.
I think it should be `MSIsensor` according to their GitHub repo, and I prefer having `MSIsensor` as the output directory, as it's the name of the tool.

**Comment by MaxUlysse** [comment]
I'll push to dockerhub a new dev container containing msisensor and relaunch the tests as well

File: docs/output.md (hunk 8 above)
Comment (davidmasp) [reply docs/output.md]: Sorry i must have delete it, it’s mismatch repair

File: main.nf (hunk 10 above)
Comment (davidmasp) [reply main.nf]: Mmmh but there’s 2 dirs caled msisensor. The first lne was equivalent to variant calling. We can put it there aswell I suppose.

Comment (MaxUlysse) [reply main.nf]: Oooh, I see, sorry about that, let me correct my suggestion then

File: docs/output.md (hunk 8 above)
Comment (MaxUlysse) [inline docs/output.md]:
```suggestion
slippage which would be corrected under normal mismatch repair (MMR) conditions. It requires
```

**Comment by davidmasp** [comment]
okay great! the MSI directory was to match the VariantCalling directory. As strelka is inside the variantcalling directory for instance. Does make sense? maybe the msisensor could go there too idk.

**Comment by MaxUlysse** [comment]
Oh, yes, sorry, I don't have my whole head on the code today.
ASCAT results are in the VariantCalling directory as well, so I'll put MSIsensor results there as well for now

File: docs/output.md [hunk 12]
```diff
@@ -424,6 +426,36 @@ For a Tumor/Normal pair only:
...
+status of a tumor scaning the length of the microsatellite regions. An altered
+distribution of  microsatellite length is associated to a missed replication
+slippage which would be corrected under normal mismatch repair (MMR) conditions. It requires
+a normal sample for each tumour to differentiate the somatic and germline
+cases.
+
+For further reading see the [MSIsensor paper](https://www.ncbi.nlm.nih.gov/pubmed/24371154).
+
+For a Tumor/Normal pair only:
+**Output directory: `results/[TUMORSAMPLE]_vs_[NORMALSAMPLE]/MSIsensor`**
```
Comment (MaxUlysse) [inline docs/output.md]:
```suggestion
**Output directory: `results/VariantCalling/[TUMORSAMPLE]_vs_[NORMALSAMPLE]/MSIsensor`**
```

File: main.nf (hunk 10 above)
Comment (davidmasp) [reply main.nf]: what I meant below is that all the other tools have their own nesting dir like `publishDir "${params.outdir}/VariantCalling` etc. That's why I originally wrote only MSI. Shouldn't we put msisensor in one of those?

File: main.nf [hunk 13]
```diff
@@ -2605,6 +2606,64 @@ process StrelkaBP {
...
+// Score the normal vs somatic pair of bams
+
+process msisensor {
+    label 'cpus_4'
+    label 'memory_max'
+    // memory '10 GB'
+
+    tag {idSampleTumor + "_vs_" + idSampleNormal}
+
+    publishDir "${params.outdir}/${idSampleTumor}_vs_${idSampleNormal}/MSIsensor", mode: params.publishDirMode
```
Comment (MaxUlysse) [inline main.nf]:
```suggestion
    publishDir "${params.outdir}/VariantCalling/${idSampleTumor}_vs_${idSampleNormal}/MSIsensor", mode: params.publishDirMode
```

File: main.nf (hunk 10 above)
Comment (davidmasp) [reply main.nf]: okok solved

**Comment by davidmasp** [comment]
Awesome! now I think all is good no?

**Comment by MaxUlysse** [comment]
> Awesome! now I think all is good no?
Yes!!!
Just waiting for the CI results and we're good :+1:

**Comment by davidmasp** [comment]
nice!

**Pull request merged**